Jenkinsfile

```diff
@@ -294,6 +306,49 @@ def runCompilerPluginTest() {
     }
 }
 
+def testWithServer(dir, tasks) {
+    // Work-around for https://github.com/docker/docker-credential-helpers/issues/82
+    withCredentials([
+            [$class: 'StringBinding', credentialsId: 'realm-kotlin-ci-password', variable: 'PASSWORD'],
+    ]) {
+        sh "security -v unlock-keychain -p $PASSWORD"
+    }
+
+    // Prepare Docker containers with MongoDB Realm Test Server infrastructure for
+    // integration tests.
+    // TODO: How much of this logic can be moved to start_server.sh for shared logic with local testing.
+    def props = readProperties file: 'dependencies.list'
+    echo "Version in dependencies.list: ${props.MONGODB_REALM_SERVER}"
+    def mdbRealmImage = docker.image("docker.pkg.github.com/realm/ci/mongodb-realm-test-server:${props.MONGODB_REALM_SERVER}")
+    docker.withRegistry('https://docker.pkg.github.com', 'github-packages-token') {
+      mdbRealmImage.pull()
+    }
+    def commandServerEnv = docker.build 'mongodb-realm-command-server', "tools/sync_test_server"
+    def tempDir = runCommand('mktemp -d -t app_config.XXXXXXXXXX')
+    sh "tools/sync_test_server/app_config_generator.sh ${tempDir} tools/sync_test_server/app_template testapp1 testapp2"
+
+    sh "docker network create ${dockerNetworkId}"
+    mongoDbRealmContainer = mdbRealmImage.run("--rm -i -t -d --network ${dockerNetworkId} -v$tempDir:/apps -p9090:9090 -p8888:8888 -p26000:26000")
+    mongoDbRealmCommandServerContainer = commandServerEnv.run("--rm -i -t -d --network container:${mongoDbRealmContainer.id} -v$tempDir:/apps")
+    sh "timeout 60 sh -c \"while [[ ! -f $tempDir/testapp1/app_id || ! -f $tempDir/testapp2/app_id ]]; do echo 'Waiting for server to start'; sleep 1; done\""
+
+    try {
+        tasks.each { task ->
+            testAndCollect(dir, task)
+        }
+    } finally {
+        // We assume that creating these containers and the docker network can be considered an atomic operation.
+        if (mongoDbRealmContainer != null && mongoDbRealmCommandServerContainer != null) {
```

Comment:
Shouldn't this be `||` and shouldn't the `image.run(...)` part above also be in the try-finally block?

Reply:
I'm using `&&` as it otherwise becomes very complicated to reason about storing server logs and shutting down things. It is true that this logic breaks down if CI creates the first container but fails to create the 2nd. However, I never noticed this happening in practice, so it didn't seem worthwhile to complicate this code for that.